### ISSUE TITLE
Z-Wave: Missing status update callback

### ISSUE THREAD

**Issue by maggu2810** [openhab/openhab-addons#819]
After worked around #815 by #818 I tried to add the Fibary Motion Sensor.

``` text
13:18:33.774 ERROR o.e.s.c.t.i.ThingManager[:711] - Exception occured during notification of thing 'zwave:device:e96123a4:node4' about bridge initialization at 'null': Could not update status, because callback is missing
java.lang.IllegalStateException: Could not update status, because callback is missing
    at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:307)
    at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:334)
    at org.openhab.binding.zwave.handler.ZWaveThingHandler.bridgeStatusChanged(ZWaveThingHandler.java:408)
    at org.openhab.binding.zwave.handler.ZWaveThingHandler.bridgeHandlerInitialized(ZWaveThingHandler.java:383)
    at org.eclipse.smarthome.core.thing.internal.ThingManager$8.run(ThingManager.java:709)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
```

https://gist.github.com/maggu2810/d293b459a2c9a6e55c234b3a19855ba7#file-gistfile1-txt-L538

**Comment by maggu2810** [comment]
Just to add that the device "FGMS001 Motion Sensor" is working but we should have a look at this error.

**Comment by cdjackson** [comment]
I saw this last weekend too - not sure why it was happening, but potentially to do with updating configuration in the initialize method (which I assume I’m allowed to do?!?) - at least that’s when it seems to have started happening - after I added this functionality. I started to trace this through, but haven’t seen it for a while - I’m not convinced it’s an issue in Z-Wave directly though...

**Comment by cdjackson** [comment]
Possible duplicate of #814

**Comment by cdjackson** [comment]
Duplicate of https://github.com/cdjackson/HABmin2/issues/169

From this log... I'm not sure how to debug this as I can't duplicate. @maggu2810 does this still happen on your system? 

```
2016-05-13 23:18:49.434 [INFO ] [smarthome.event.InboxRemovedEvent   ] - Discovery Result with UID 'zwave:device:154acc671d8:node3' has been removed.
2016-05-13 23:18:49.457 [INFO ] [smarthome.event.ThingAddedEvent     ] - Thing 'zwave:device:154acc671d8:node3' has been added.
2016-05-13 23:18:49.478 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:device:154acc671d8:node3' changed from UNINITIALIZED to INITIALIZING
2016-05-13 23:18:49.507 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while initializing handler of thing 'zwave:device:154acc671d8:node3': java.lang.IllegalStateException: Could not update status, because callback is missing
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Could not update status, because callback is missing
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0_91]
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0_91]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:179)[96:org.eclipse.smarthome.core:0.8.0.201605112054]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)[96:org.eclipse.smarthome.core:0.8.0.201605112054]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)[96:org.eclipse.smarthome.core:0.8.0.201605112054]
	at org.eclipse.smarthome.core.thing.internal.ThingManager$7.run(ThingManager.java:696)[102:org.eclipse.smarthome.core.thing:0.8.0.201605112054]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_91]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_91]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)[:1.8.0_91]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)[:1.8.0_91]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_91]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_91]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_91]
Caused by: java.lang.IllegalStateException: Could not update status, because callback is missing
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:307)[102:org.eclipse.smarthome.core.thing:0.8.0.201605112054]
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:334)[102:org.eclipse.smarthome.core.thing:0.8.0.201605112054]
	at org.openhab.binding.zwave.handler.ZWaveThingHandler.initialize(ZWaveThingHandler.java:130)[196:org.openhab.binding.zwave:2.0.0.201605130102]
	at org.eclipse.smarthome.core.thing.internal.ThingManager$7$1.call(ThingManager.java:699)[102:org.eclipse.smarthome.core.thing:0.8.0.201605112054]
	at org.eclipse.smarthome.core.thing.internal.ThingManager$7$1.call(ThingManager.java:1)[102:org.eclipse.smarthome.core.thing:0.8.0.201605112054]
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)[96:org.eclipse.smarthome.core:0.8.0.201605112054]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_91]
	... 3 more
2016-05-13 23:18:49.522 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:device:154acc671d8:node3' changed from INITIALIZING to UNINITIALIZED (HANDLER_MISSING_ERROR)
2016-05-13 23:18:49.532 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:device:154acc671d8:node3' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to UNINITIALIZED (HANDLER_INITIALIZING_ERROR): Could not update status, because callback is missing
```

**Comment by mhilbush** [comment]
Let me know if there's anything I can do to help debug this.  If there's a possibility that it has to do with updating config in the initialize method, as you state above, is it possible for me to run a version prior to you making that change?

**Comment by maggu2810** [comment]
@cdjackson I suspend using the OH2 binding on the system the error occurred. I need the manu. prop. support there.
But I can try to test it on another one next week.

**Comment by gdombiak** [comment]
I see this happening using 2.0.0.201605280102. I know how to reproduce it consistently if you need more info. It happens with some type of zwave devices. In my case 'DSC06 Smart Energy Switch' and '45605 Duplex receptacle'.

Here you can see the stack trace and also the next steps in the flow where things get corrected immediately.

> 2016-05-30 10:09:41.483 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while initializing handler of thing 'zwave:device:d4f40fe7:node13': java.lang.IllegalStateException: Could not update configuration, because callback is missing
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Could not update configuration, because callback is missing
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0_65]
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0_65]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:179)[96:org.eclipse.smarthome.core:0.8.0.201605191852]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)[96:org.eclipse.smarthome.core:0.8.0.201605191852]
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)[96:org.eclipse.smarthome.core:0.8.0.201605191852]
	at org.eclipse.smarthome.core.thing.internal.ThingManager$7.run(ThingManager.java:684)[102:org.eclipse.smarthome.core.thing:0.8.0.201605191852]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_65]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_65]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)[:1.8.0_65]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)[:1.8.0_65]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_65]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_65]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_65]
Caused by: java.lang.IllegalStateException: Could not update configuration, because callback is missing
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateConfiguration(BaseThingHandler.java:419)[102:org.eclipse.smarthome.core.thing:0.8.0.201605191852]
	at org.eclipse.smarthome.core.thing.binding.ConfigStatusThingHandler.updateConfiguration(ConfigStatusThingHandler.java:64)[102:org.eclipse.smarthome.core.thing:0.8.0.201605191852]
	at org.openhab.binding.zwave.handler.ZWaveThingHandler.bridgeStatusChanged(ZWaveThingHandler.java:513)[188:org.openhab.binding.zwave:2.0.0.201605280102]
	at org.openhab.binding.zwave.handler.ZWaveThingHandler.initialize(ZWaveThingHandler.java:140)[188:org.openhab.binding.zwave:2.0.0.201605280102]
	at org.eclipse.smarthome.core.thing.internal.ThingManager$7$1.call(ThingManager.java:687)[102:org.eclipse.smarthome.core.thing:0.8.0.201605191852]
	at org.eclipse.smarthome.core.thing.internal.ThingManager$7$1.call(ThingManager.java:1)[102:org.eclipse.smarthome.core.thing:0.8.0.201605191852]
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)[96:org.eclipse.smarthome.core:0.8.0.201605191852]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_65]
	... 3 more
2016-05-30 10:09:41.492 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:device:d4f40fe7:node13' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to UNINITIALIZED (HANDLER_INITIALIZING_ERROR): Could not update configuration, because callback is missing
2016-05-30 10:09:41.500 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:device:d4f40fe7:node13' changed from OFFLINE to ONLINE
2016-05-30 10:09:41.765 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:device:d4f40fe7:node13' changed from UNINITIALIZED (HANDLER_INITIALIZING_ERROR): Could not update configuration, because callback is missing to INITIALIZING
2016-05-30 10:09:41.789 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:device:d4f40fe7:node13' changed from INITIALIZING to OFFLINE
2016-05-30 10:09:41.796 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:device:d4f40fe7:node13' changed from OFFLINE to ONLINE

Thanks,
Gaston

**Comment by cdjackson** [comment]
@kaikreuzer @maggu2810 I think that this issue might be related to the service startup problem (ie potentially the circular reference problem). I'm now seeing this error showing up on a number of bindings. Looking at the code, I can't really see how the callback can't be set if the service registration is working since it is set directly in the service setter if I remember correctly.

```
21:43:07.634 [DEBUG] [org.openhab.binding.zwave           ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandler}={thing.type=zwave:tkb_tz88_00_000, thing.id=zwave:device:1531045fdae:node13, service.id=394, service.bundleid=203, service.scope=singleton} - org.openhab.binding.zwave
21:43:07.633 [DEBUG] [ding.zwave.handler.ZWaveThingHandler] - Initializing ZWave thing handler.
21:43:07.643 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while initializing handler of thing 'zwave:device:1531045fdae:node4': java.lang.IllegalStateException: Could not update status, because callback is missing
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Could not update status, because callback is missing
```

**Comment by maggu2810** [comment]
@cdjackson If you want to give it a try, you can remove the lazy bundle activation policy from the model bundles' manifest, rebuild ESH and change the model bundles in your runtime.
If the error is solved, then there is a big change that it is a duplicate of eclipse/smarthome#870 respectively eclipse/smarthome#947.

**Comment by kaikreuzer** [comment]
FYI, I am just testing https://github.com/eclipse/smarthome/pull/947 and the runtime as well as the designer seem to work well - I am just trying to analyze the failing test case now. Hope that this could finally solve the issues then.

**Comment by kaikreuzer** [comment]
FTR: the latest build [395](https://openhab.ci.cloudbees.com/job/openHAB-Distribution/395/) includes the fix https://github.com/eclipse/smarthome/pull/1810. With that, the circular dependency should be solves, but I am just seeing sometimes some other weird log errors - so I am not sure, if this is 100% solved already, but please feel free to test and report back!

**Comment by kaikreuzer** [comment]
>  I am just seeing sometimes some other weird log errors

FTR: I think the startup problems are finally solved with https://github.com/eclipse/smarthome/pull/1828.

So if this issue was due to any of these, it should be gone now. If not, it is something separate that needs to be analyzed.

**Comment by cdjackson** [comment]
Thanks - so is this flowed through to the latest OH2? I guess so.

My ZWave is running this morning with the latest snapshot, although I’ve not tried to stop/start bindings to see if it works every time.

**Comment by kaikreuzer** [comment]
> Thanks - so is this flowed through to the latest OH2? I guess so.

Yes, see https://community.openhab.org/t/circular-dependencies-startup-issues-resolved/12266.

**Comment by cdjackson** [comment]
@kai Unfortunately this doesn't seem to be fixed. I'm seeing this on a different binding at the moment (currently these same errors on ZWave have not appeared since I loaded the updated binding)  -:

```
17:14:50.306 [DEBUG] [org.openhab.binding.meteostick      ] - BundleEvent STARTING - org.openhab.binding.meteostick
17:14:50.318 [DEBUG] [ck.internal.meteostickHandlerFactory] - MeteoStick thing factory: createHandler meteostick:meteostick_bridge of type meteostick:meteostick_bridge:155d047f32b
17:14:50.325 [DEBUG] [org.openhab.binding.meteostick      ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandler}={thing.type=meteostick:meteostick_bridge, thing.id=meteostick:meteostick_bridge:155d047f32b, service.id=331, service.bundleid=206, service.scope=singleton} - org.openhab.binding.meteostick
17:14:50.326 [DEBUG] [ck.internal.meteostickHandlerFactory] - MeteoStick thing factory: createHandler meteostick:meteostick_davis_iss of type meteostick:meteostick_davis_iss:155d0484351
17:14:50.333 [DEBUG] [tick.handler.meteostickBridgeHandler] - Initializing MeteoStick Bridge handler.
17:14:50.334 [DEBUG] [org.openhab.binding.meteostick      ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandler}={thing.type=meteostick:meteostick_davis_iss, thing.id=meteostick:meteostick_davis_iss:155d0484351, service.id=332, service.bundleid=206, service.scope=singleton} - org.openhab.binding.meteostick
17:14:50.335 [INFO ] [tick.handler.meteostickBridgeHandler] - MeteoStick Connecting to serial port /dev/ttyUSB0
17:14:50.335 [DEBUG] [org.openhab.binding.meteostick      ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory}={component.name=org.openhab.binding.meteostick.internal.meteostickHandlerFactory, component.id=195, service.id=330, service.bundleid=206, service.scope=bundle} - org.openhab.binding.meteostick
17:14:50.336 [DEBUG] [tick.handler.meteostickSensorHandler] - Initializing MeteoStick handler.
17:14:50.339 [DEBUG] [tick.handler.meteostickSensorHandler] - Initializing MeteoStick handler - Channel 1.
17:14:50.343 [DEBUG] [org.openhab.binding.meteostick      ] - BundleEvent STARTED - org.openhab.binding.meteostick
17:15:17.801 [DEBUG] [eteostickBridgeHandler$ReceiveThread] - MeteoStick received: B 20.7 1008.80 0.0%
17:15:17.802 [ERROR] [eteostickBridgeHandler$ReceiveThread] - Exception during MeteoStick receive thread
java.lang.IllegalStateException: Could not update state, because callback is missing
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:258)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.openhab.binding.meteostick.handler.meteostickBridgeHandler.access$1(meteostickBridgeHandler.java:1)[203:org.openhab.binding.meteostick:2.0.0.201607031305]
        at org.openhab.binding.meteostick.handler.meteostickBridgeHandler$ReceiveThread.run(meteostickBridgeHandler.java:232)[203:org.openhab.binding.meteostick:2.0.0.201607031305]
```

Looking at this, it registers the bridge, then registers the thing. The bridge handler sets the status to offline. Then the bridge receives data and tries to update the status to online and it throws the no callback exception.

I assume that in the ```initialize``` method of the handler where it sets the status to offline, that the callback must be present, but shortly after, it seems not to be.

**Comment by maggu2810** [comment]
> I assume that in the initialize method of the handler where it sets the status to offline, that the callback must be present, but shortly after, it seems not to be.

You can give it a try and overwrite the method, perhaps you can identify at which time the callback is set / unset.

```java
    @Override
    public void setCallback(ThingHandlerCallback thingHandlerCallback) {
        logger.debug("Callback is set to: {}.", thingHandlerCallback, new RuntimeException("log stacktrace"));
        super.setCallback(thingHandlerCallback);
    }
```

**Comment by cdjackson** [comment]
Nice idea, but I don't think it helps -:

```
20:39:09.176 [INFO ] [org.apache.felix.fileinstall        ] - Installing bundle org.openhab.binding.meteostick;singleton:=true / 2.0.0.201607092138
20:39:09.199 [DEBUG] [org.openhab.binding.meteostick      ] - BundleEvent INSTALLED - org.openhab.binding.meteostick
20:39:09.222 [DEBUG] [org.openhab.binding.meteostick      ] - BundleEvent RESOLVED - org.openhab.binding.meteostick
20:39:09.223 [DEBUG] [org.openhab.binding.meteostick      ] - BundleEvent STARTING - org.openhab.binding.meteostick
20:39:09.237 [DEBUG] [ck.internal.meteostickHandlerFactory] - MeteoStick thing factory: createHandler meteostick:meteostick_bridge of type meteostick:meteostick_bridge:155d047f32b
20:39:09.242 [DEBUG] [tick.handler.meteostickBridgeHandler] - Callback is set to: org.eclipse.smarthome.core.thing.internal.ThingManager$1@20e71d3.
java.lang.RuntimeException: log stacktrace
        at org.openhab.binding.meteostick.handler.meteostickBridgeHandler.setCallback(meteostickBridgeHandler.java:70)[208:org.openhab.binding.meteostick:2.0.0.201607092138]
        at org.eclipse.smarthome.core.thing.internal.ThingManager.handlerAdded(ThingManager.java:348)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$ThingHandlerTracker.addingService(ThingManager.java:117)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$ThingHandlerTracker.addingService(ThingManager.java:1)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:941)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:870)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)[org.osgi.core-6.0.0.jar:]
        at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:901)[org.osgi.core-6.0.0.jar:]
        at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:914)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:127)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:225)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:464)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:482)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerThingHandlerAsService(BaseThingHandlerFactory.java:167)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:145)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:609)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:1)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:177)[97:org.eclipse.smarthome.core:0.9.0.201607082129]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_60]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_60]
20:39:09.244 [DEBUG] [org.openhab.binding.meteostick      ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandler}={thing.type=meteostick:meteostick_bridge, thing.id=meteostick:meteostick_bridge:155d047f32b, service.id=350, service.bundleid=208, service.scope=singleton} - org.openhab.binding.meteostick
20:39:09.244 [DEBUG] [tick.handler.meteostickBridgeHandler] - Initializing MeteoStick Bridge handler.
20:39:09.245 [DEBUG] [ck.internal.meteostickHandlerFactory] - MeteoStick thing factory: createHandler meteostick:meteostick_davis_iss of type meteostick:meteostick_davis_iss:155d0484351
20:39:09.249 [INFO ] [tick.handler.meteostickBridgeHandler] - MeteoStick Connecting to serial port /dev/ttyUSB0
20:39:09.249 [DEBUG] [org.openhab.binding.meteostick      ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandler}={thing.type=meteostick:meteostick_davis_iss, thing.id=meteostick:meteostick_davis_iss:155d0484351, service.id=351, service.bundleid=208, service.scope=singleton} - org.openhab.binding.meteostick
20:39:09.251 [DEBUG] [tick.handler.meteostickSensorHandler] - Initializing MeteoStick handler.
20:39:09.252 [DEBUG] [org.openhab.binding.meteostick      ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory}={component.name=org.openhab.binding.meteostick.internal.meteostickHandlerFactory, component.id=199, service.id=349, service.bundleid=208, service.scope=bundle} - org.openhab.binding.meteostick
20:39:09.252 [DEBUG] [tick.handler.meteostickSensorHandler] - Initializing MeteoStick handler - Channel 1.
20:39:09.257 [INFO ] [g.apache.aries.spifly.dynamic.bundle] - Bundle Considered for SPI providers: org.openhab.binding.meteostick
20:39:09.257 [INFO ] [g.apache.aries.spifly.dynamic.bundle] - No 'SPI-Provider' Manifest header. Skipping bundle: org.openhab.binding.meteostick
20:39:09.257 [DEBUG] [org.openhab.binding.meteostick      ] - BundleEvent STARTED - org.openhab.binding.meteostick
20:39:09.257 [INFO ] [org.apache.felix.fileinstall        ] - Started bundle: file:/volume1/openHAB-2/addons/org.openhab.binding.meteostick_2.0.0.201607092138.jar
20:39:22.163 [DEBUG] [eteostickBridgeHandler$ReceiveThread] - MeteoStick received: B 20.7 1007.34 0.0%
20:39:22.286 [ERROR] [eteostickBridgeHandler$ReceiveThread] - Exception during MeteoStick receive thread
java.lang.IllegalStateException: Could not update state, because callback is missing
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:258)[103:org.eclipse.smarthome.core.thing:0.9.0.201607082129]
        at org.openhab.binding.meteostick.handler.meteostickBridgeHandler.access$1(meteostickBridgeHandler.java:1)[203:org.openhab.binding.meteostick:2.0.0.201607031305]
        at org.openhab.binding.meteostick.handler.meteostickBridgeHandler$ReceiveThread.run(meteostickBridgeHandler.java:232)[203:org.openhab.binding.meteostick:2.0.0.201607031305]
```

It looks like the callback gets set ok, and only once!

Any thoughts...

**Comment by maggu2810** [comment]
Just another idea:
Can you add

```java
logger.debug("identity: {}", System.identifyHashCode(this));
```

to some functions of your bridge handler?
e.g. constructor, initialize, setCallback, etc.
Just to ensure that there are really only one object, and the one that misses the callback is not the same one 'setCalback' has been called.

**Comment by cdjackson** [comment]
Given that the problem area is inside an internal thread, “this" doesn’t work (i.e. it picks up on the thread I would guess).

**Comment by kaikreuzer** [comment]
Where is the problem using the debugger? You should clearly see which handler instance gets the callback and why it is missing when being used.

**Comment by cdjackson** [comment]
It works fine in the debugger.

**Comment by kaikreuzer** [comment]
Always or only when you step through the code?

**Comment by cdjackson** [comment]
Always. I don’t see any issues with any of the startup problems in the IDE - only in my “online” system.

**Comment by kaikreuzer** [comment]
Then please use remote debugging of your online system!

**Comment by sbussweiler** [comment]
@maggu2810 Regarding your logs from the initial post of this issue...

```
13:18:33.653 DEBUG o.e.s.c.t.i.ThingManager[:449] - Thing 'zwave:device:e96123a4:node4' is tracked by ThingManager.
13:18:33.655 DEBUG o.e.s.c.t.i.ThingManager[:580] - Calling registerHandler handler for thing 'zwave:device:e96123a4:node4' at 'org.openhab.binding.zwave.internal.ZWaveHandlerFactory@661f3622'.
13:18:33.655 INFO  s.event.InboxRemovedEvent[:43] - Discovery Result with UID 'zwave:device:e96123a4:node4' has been removed.
13:18:33.656 DEBUG o.e.s.c.t.b.BaseThingHandlerFactory[:44] - Creating thing zwave:device:e96123a4:node4
13:18:33.656 INFO  s.event.ThingAddedEvent[:43] - Thing 'zwave:device:e96123a4:node4' has been added.
13:18:33.662 DEBUG o.e.s.c.t.i.ThingManager[:101] - Thing handler for thing 'zwave:device:e96123a4:node4' added.
13:18:33.663 DEBUG o.e.s.c.t.i.ThingManager[:319] - Assigning handler for thing 'zwave:device:e96123a4:node4'.
13:18:33.663 INFO  s.event.ThingStatusInfoEvent[:43] - 'zwave:device:e96123a4:node4' updated: INITIALIZING
13:18:33.665 INFO  s.e.ThingStatusInfoChangedEvent[:43] - 'zwave:device:e96123a4:node4' changed from UNINITIALIZED to INITIALIZING
13:18:33.665 DEBUG o.e.s.c.t.i.ThingManager[:657] - Calling initialize handler for thing 'zwave:device:e96123a4:node4' at 'org.openhab.binding.zwave.handler.ZWaveThingHandler@7fb8c01f'.
13:18:33.673 DEBUG o.e.s.c.c.r.AbstractManagedProvider[:62] - Added new element zwave:device:e96123a4:node4 to ManagedThingProvider.
13:18:33.677 WARN  o.o.b.z.h.ZWaveThingHandler[:140] - NODE 4: Initialising Thing Node...
13:18:33.679 INFO  s.event.ThingStatusInfoEvent[:43] - 'zwave:device:e96123a4:node4' updated: OFFLINE
13:18:33.679 DEBUG o.e.s.c.t.i.ThingManager[:720] - Calling unregisterHandler handler for thing 'zwave:device:e96123a4:node4' at 'org.openhab.binding.zwave.internal.ZWaveHandlerFactory@661f3622'.
13:18:33.680 INFO  s.e.ThingStatusInfoChangedEvent[:43] - 'zwave:device:e96123a4:node4' changed from INITIALIZING to OFFLINE
13:18:33.683 INFO  s.event.ThingStatusInfoEvent[:43] - 'zwave:device:e96123a4:node4' updated: ONLINE
13:18:33.684 INFO  s.e.ThingStatusInfoChangedEvent[:43] - 'zwave:device:e96123a4:node4' changed from OFFLINE to ONLINE
13:18:33.686 DEBUG o.e.s.c.t.i.ThingManager[:119] - Thing handler for thing 'zwave:device:e96123a4:node4' removed.
13:18:33.686 DEBUG o.e.s.c.t.i.ThingManager[:334] - Unassigning handler for thing 'zwave:device:e96123a4:node4' and setting status to UNINITIALIZED.
13:18:33.688 INFO  s.event.ThingStatusInfoEvent[:43] - 'zwave:device:e96123a4:node4' updated: UNINITIALIZED (HANDLER_MISSING_ERROR)
13:18:33.689 INFO  s.e.ThingStatusInfoChangedEvent[:43] - 'zwave:device:e96123a4:node4' changed from ONLINE to UNINITIALIZED (HANDLER_MISSING_ERROR)
13:18:33.716 DEBUG o.e.s.c.c.r.AbstractManagedProvider[:62] - Added new element zwave_device_e96123a4_node4 to ManagedItemProvider.
13:18:33.717 INFO  smarthome.event.ItemAddedEvent[:43] - Item 'zwave_device_e96123a4_node4' has been added.
13:18:33.754 DEBUG o.e.s.c.t.i.ThingManager[:529] - Cannot notify handler about updated thing zwave:device, because thing is not initialized (must be in status ONLINE or OFFLINE).
13:18:33.755 DEBUG o.e.s.c.c.r.AbstractManagedProvider[:135] - Updated element zwave:device:e96123a4:node4 in ManagedThingProvider.
13:18:33.755 DEBUG o.e.s.c.t.i.ThingManager[:736] - Calling dispose handler for thing 'zwave:device:e96123a4:node4' at 'org.openhab.binding.zwave.handler.ZWaveThingHandler@7fb8c01f'.
13:18:33.756 INFO  s.event.ThingUpdatedEvent[:43] - Thing 'zwave:device:e96123a4:node4' has been updated.
13:18:33.768 INFO  s.event.ConfigStatusInfoEvent[:43] - org.eclipse.smarthome.config.core.status.events.ConfigStatusInfoEvent@1ce7a7
13:18:33.774 ERROR o.e.s.c.t.i.ThingManager[:711] - Exception occured during notification of thing 'zwave:device:e96123a4:node4' about bridge initialization at 'null': Could not update status, because callback is missing
java.lang.IllegalStateException: Could not update status, because callback is missing
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:307)
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:334)
	at org.openhab.binding.zwave.handler.ZWaveThingHandler.bridgeStatusChanged(ZWaveThingHandler.java:408)
	at org.openhab.binding.zwave.handler.ZWaveThingHandler.bridgeHandlerInitialized(ZWaveThingHandler.java:383)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$8.run(ThingManager.java:709)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
```

... the thing handler will be disposed, the callback is set to null. In the same moment the ThingManager tries to notify asynchronously the thing about an initialized bridge, caused by a previous status change. But the handler could not update the status, because the callback is missing now.

I thing an implementation of the new bridge/thing lifecycle (see https://github.com/eclipse/smarthome/issues/1479) would fix that issue. Is this solution fine for you, @maggu2810?

**Comment by maggu2810** [comment]
@sbussweiler Sure :wink: I assume it will improve the whole life-cycle situation sensibly.

**Comment by sbussweiler** [comment]
That's the plan ;-)

**Comment by cdjackson** [comment]
I hope / think that this is fixed with the changes to the startup - certainly this issue has gone away on other bindings after this change.

I'll close this - feel free to re-open if it still exists.

**Comment by robkamp** [comment]
From my logfile today
```
2017-09-26 14:19:38.900 [ERROR] [ve.internal.protocol.ZWaveController] - Exception during ZWave thread: Input 2. {}
java.lang.IllegalStateException: Could not update state, because callback is missing
```
I run OpenHAB 2.1 with z-wave 2.1.0 binding.

**Comment by maxlein** [comment]
Same with me here:
Exception with ZWave Plus device PST02A 
Openhab version 2.1.0:

`017-10-01 14:15:41.711 [DEBUG] [ve.internal.protocol.ZWaveController] - Receive queue TAKE: Length=0
2017-10-01 14:15:41.714 [DEBUG] [wave.internal.protocol.SerialMessage] - Assembled message buffer = 01 0C 00 04 00 0A 06 31 05 01 0A 00 4B 8F 
2017-10-01 14:15:41.717 [DEBUG] [ve.internal.protocol.ZWaveController] - Process Message = 01 0C 00 04 00 0A 06 31 05 01 0A 00 4B 8F 
2017-10-01 14:15:41.721 [DEBUG] [ve.internal.protocol.ZWaveController] - Message: class=ApplicationCommandHandler[0x04], type=Request[0x00], priority=High, dest=255, callback=0, payload=00 0A 06 31 05 01 0A 00 4B 
2017-10-01 14:15:41.723 [DEBUG] [ssage.ApplicationCommandMessageClass] - NODE 10: Application Command Request (ALIVE:DONE)
2017-10-01 14:15:41.725 [DEBUG] [alization.ZWaveNodeInitStageAdvancer] - NODE 10: Starting initialisation from DONE
2017-10-01 14:15:41.727 [DEBUG] [ve.internal.protocol.ZWaveController] - Event Listener org.openhab.binding.zwave.internal.protocol.initialization.ZWaveNodeInitStageAdvancer@1d4b448 already registered
2017-10-01 14:15:41.738 [DEBUG] [ssage.ApplicationCommandMessageClass] - NODE 10: Incoming command class SENSOR_MULTILEVEL
2017-10-01 14:15:41.740 [DEBUG] [ss.ZWaveMultiLevelSensorCommandClass] - NODE 10: Received COMMAND_CLASS_SENSOR_MULTILEVEL command V5
2017-10-01 14:15:41.742 [DEBUG] [ss.ZWaveMultiLevelSensorCommandClass] - NODE 10: Sensor Multi Level REPORT received
2017-10-01 14:15:41.757 [DEBUG] [ss.ZWaveMultiLevelSensorCommandClass] - NODE 10: Sensor Type = Temperature(1), Scale = 1
2017-10-01 14:15:41.759 [DEBUG] [ss.ZWaveMultiLevelSensorCommandClass] - NODE 10: Sensor Value = 75
2017-10-01 14:15:41.760 [DEBUG] [ve.internal.protocol.ZWaveController] - Notifying event listeners: ZWaveMultiLevelSensorValueEvent
2017-10-01 14:15:41.761 [DEBUG] [ding.zwave.handler.ZWaveThingHandler] - NODE 10: Got an event from Z-Wave network: ZWaveMultiLevelSensorValueEvent
2017-10-01 14:15:41.763 [DEBUG] [ding.zwave.handler.ZWaveThingHandler] - NODE 10: Got a value event from Z-Wave network, endpoint = 0, command class = SENSOR_MULTILEVEL, value = 75
2017-10-01 14:15:41.764 [DEBUG] [erter.ZWaveMultiLevelSensorConverter] - NODE 10: Sensor is reporting scale 1, requiring conversion to 0. Value is now 75.
2017-10-01 14:15:41.766 [DEBUG] [converter.ZWaveCommandClassConverter] - Converted temperature from 75F to 23.9C
2017-10-01 14:15:41.767 [DEBUG] [ding.zwave.handler.ZWaveThingHandler] - NODE 10: Updating channel state zwave:device:15a670b08a0:node10:sensor_temperature to 23.9 [DecimalType]
2017-10-01 14:15:41.768 [ERROR] [ve.internal.protocol.ZWaveController] - Exception during ZWave thread: Input 2. {}
java.lang.IllegalStateException: Could not update state, because callback is missing
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:267)[105:org.eclipse.smarthome.core.thing:0.9.0.b5]
	at org.openhab.binding.zwave.handler.ZWaveThingHandler.ZWaveIncomingEvent(ZWaveThingHandler.java:1104)[189:org.openhab.binding.zwave:2.1.0]
	at org.openhab.binding.zwave.internal.protocol.ZWaveController.notifyEventListeners(ZWaveController.java:567)[189:org.openhab.binding.zwave:2.1.0]
	at org.openhab.binding.zwave.internal.protocol.commandclass.ZWaveMultiLevelSensorCommandClass.handleApplicationCommandRequest(ZWaveMultiLevelSensorCommandClass.java:150)[189:org.openhab.binding.zwave:2.1.0]
	at org.openhab.binding.zwave.internal.protocol.serialmessage.ApplicationCommandMessageClass.handleRequest(ApplicationCommandMessageClass.java:119)[189:org.openhab.binding.zwave:2.1.0]
	at org.openhab.binding.zwave.internal.protocol.ZWaveController.handleIncomingRequestMessage(ZWaveController.java:248)[189:org.openhab.binding.zwave:2.1.0]
	at org.openhab.binding.zwave.internal.protocol.ZWaveController.handleIncomingMessage(ZWaveController.java:214)[189:org.openhab.binding.zwave:2.1.0]
	at org.openhab.binding.zwave.internal.protocol.ZWaveController.access$7(ZWaveController.java:208)[189:org.openhab.binding.zwave:2.1.0]
	at org.openhab.binding.zwave.internal.protocol.ZWaveController$ZWaveInputThread.run(ZWaveController.java:1324)[189:org.openhab.binding.zwave:2.1.0]
`

**Comment by martinvw** [comment]
@cdjackson ^^

**Comment by cdjackson** [comment]
Sorry - but what does ^^ mean?

**Comment by martinvw** [comment]
I did not know whether you would receive the update given above, so I thought let's mention you and point to the message above. I thought it was more or less common 😁

**Comment by cdjackson** [comment]
Ok, I’ve not come across that before ;)

I am not following this though - I’d suggest to open something on the ZWave list - especially as this seems to have been closed 18 months ago it’s hard to track.

**Comment by robkamp** [comment]
Where can I find the ZWave list?

**Comment by martinvw** [comment]
https://github.com/openhab/org.openhab.binding.zwave/issues

**Comment by robkamp** [comment]
@martinvw thanks!
